Things we should consider for everything we ship (new and existing features) (#4601)
sourcegraph/handbook: content/departments/engineering/index.md

- [Working with issues](working-with-issues.md)
- [Our software development lifecycle (SDLC)](sdlc.md)

## What's in a feature?

> For every feature we ship, consider:

Reviewer: Is this an ordered list?

Author: No (except that putting security first was intentional).

> For every feature we ship, consider:

Reviewer 2: what does it mean to "consider" – is this a requirement?

 If it's not a requirement, how should people use this? 

If it is a requirement, can we provide staged tiers (specifically around scalability – neither batch changes nor code insights works well for 100ks+ repos right now, but obviously they still provide tons of value). I'm happy to own defining tiered scale support here on this page or somewhere else if that'd be valuable.

- **Security:** Make it secure, because our customers trust us with very sensitive code and other data, and we need to continually earn and retain that trust.
- **Reliability:** Make it stable and robust, because devs don't trust flaky or incorrect results.

> - **Speed:** Make it fast, because devs won't use something 10+ times daily if it's slow.

Reviewer 3: I really like how all statements up to this one end with a "[...], because [...]". Makes it easy for everyone to empathize with why it's important for customers. Maybe we could do it for more points?

- **Scalability:** Make it work at large scale, such as big monorepos or 100,000s of repositories.
- **Documentation:** Add and update documentation for the new feature.
- **Announcement:** Draft the eventual announcement of the feature early (in blog post form, along with a changelog entry and major docs changes), and update it as you iterate.
- **Iteration feedback:** Iterate with our existing customers and gather user feedback as early as possible.
- **Usage/value analytics:** Show how the feature is being used to users and admins, communicate the value to the customer's business, and identify opportunities for more value.

> - **Accessibility:** Make it accessible to a broad set of users (including keyboard accessibility).

Reviewer 3: When talking about accessibility, I worry that "a broad set of users" is (in many minds) satisfied even by just targeting the "common" user with no impairments (i.e., basically _not_ caring about accessibility). If we want to state this as a goal, I think we need to make that stronger
```suggestion
- **Accessibility:** Make it accessible to all users (including keyboard accessibility).
```

- **Marketing:** Update all marketing content to reflect the new feature (such as our about site, readmes, extension descriptions, slides, etc.).
- **Observability:** Help admins debug, manage, and scale the functionality.
- **Telemetry:** Gather the (minimum) data needed for us to understand how the feature is being used and improve it.
- **Backward-compatibility:** Make it smooth to upgrade from past versions and adapt usage to the newest version.
- **Forward-compatibility:** Anticipate the feature's future evolution (where possible, and without over-building today).

> - **Pricing/packaging:** Decide which plans include this feature and whether its code will be open-source or not.

Reviewer 4: I think Reliability touches on this but would like to see a specific call out to *Testing* - How we will test a feature so we understand how it behaves.

Author: Can you send a followup PR with what you'd propose here? I see testing as "how to achieve these things" rather than a goal in itself, but it still might be worth calling out specifically here.
